File: ios/joinery-kit/Sources/JoineryMailKit/ComposeSheet.swift
import SwiftUI
import PhotosUI
import UniformTypeIdentifiers
import UIKit
import JoineryKit

/// Reply / reply-all / forward / new-message compose. Deliberately lean: the
/// server is the authority on quoting, subject normalization (Re:/Fwd:),
/// threading headers, and the sending identity (for a new message, the
/// picked mailbox) — this sheet collects recipients and the new text. A
/// forward re-attaches the original server-side; new uploads (any mode)
/// attach here via Photo Library / Files, mirroring the AI chat composer's
/// attach flow (specs/implemented/inbound_email_compose_attachments.md,
/// specs/implemented/inbound_email_new_message_compose.md).
struct ComposeSheet: View {
    let api: MailAPI
    let request: ComposeRequest
    /// The viewer's granted mailboxes — only consulted in `.new` mode, for the
    /// From picker (reply/forward keep their implicit source-derived identity).
    let mailboxes: [Mailbox]
    let onSent: () -> Void

    @State private var to: String
    @State private var cc: String = ""
    @State private var subject: String
    @State private var fromAlias: Int?
    @State private var bodyText = ""
    @State private var attachments: [MailOutgoingAttachment] = []
    @State private var showPhotoPicker = false
    @State private var photoSelection: [PhotosPickerItem] = []
    @State private var showFileImporter = false
    @State private var isSending = false
    @State private var failure: String?
    @Environment(\.dismiss) private var dismiss
    @FocusState private var bodyFocused: Bool

    /// Preflight only — mirrors the server's real caps
    /// (`MailboxSender::MAX_UPLOAD_FILES/MAX_UPLOAD_BYTES/MAX_TOTAL_BYTES`); the
    /// server remains the authority and re-validates every file and the total.
    private static let maxAttachments = 10
    private static let maxAttachmentBytes = 10_485_760
    private static let maxTotalBytes = 26_214_400

    /// No `accept` filter on Files — email legitimately carries arbitrary file
    /// types, matching the server's no-allowlist policy.
    private static let documentTypes: [UTType] = [.item]

    /// Image UTTypes the server accepts as-is; anything else the Photos picker
    /// hands back (notably HEIC, the iPhone default) is transcoded to JPEG.
    private static let directImageTypes: Set<String> = [
        UTType.png.identifier, UTType.jpeg.identifier, UTType.gif.identifier,
        UTType.webP.identifier,
    ]

    init(api: MailAPI, request: ComposeRequest, mailboxes: [Mailbox] = [],
         preselectedAlias: Int? = nil, onSent: @escaping () -> Void) {
        self.api = api
        self.request = request
        self.mailboxes = mailboxes
        self.onSent = onSent

        switch request.mode {
        case .reply, .replyAll:
            if let source = request.source {
                // Replying to your own outbound message goes back to its
                // recipient; otherwise to the sender.
                let target = source.isOutbound ? source.recipient : source.sender
                _to = State(initialValue: MailDisplay.address(target))
                _subject = State(initialValue: Self.prefixed(source.subject, "Re:"))
            } else {
                _to = State(initialValue: "")
                _subject = State(initialValue: "")
            }
            _fromAlias = State(initialValue: nil)
        case .forward:
            _to = State(initialValue: "")
            _subject = State(initialValue: Self.prefixed(request.source?.subject ?? "", "Fwd:"))
            _fromAlias = State(initialValue: nil)
        case .new:
            _to = State(initialValue: "")
            _subject = State(initialValue: "")
            _fromAlias = State(initialValue: preselectedAlias ?? mailboxes.first?.aliasID)
        }
    }

    private static func prefixed(_ subject: String, _ prefix: String) -> String {
        let trimmed = subject.trimmingCharacters(in: .whitespaces)
        if trimmed.lowercased().hasPrefix(prefix.lowercased()) { return trimmed }
        return "\(prefix) \(trimmed)"
    }

    private var title: String {
        switch request.mode {
        case .reply: return "Reply"
        case .replyAll: return "Reply all"
        case .forward: return "Forward"
        case .new: return "New message"
        }
    }

    /// No footer for a new message — there is nothing quoted or forwarded.
    private var footerText: String? {
        switch request.mode {
        case .forward: return "The forwarded message and its attachments are included below your text."
        case .new: return nil
        case .reply, .replyAll: return "The original message is quoted below your text."
        }
    }

    var body: some View {
        NavigationStack {
            Form {
                if request.mode == .new {
                    Section {
                        Picker("From", selection: $fromAlias) {
                            ForEach(mailboxes) { box in
                                Text(box.address).tag(Optional(box.aliasID))
                            }
                        }
                        .accessibilityIdentifier("mail_compose_from")
                    }
                }
                Section {
                    TextField("To", text: $to)
                        .keyboardType(.emailAddress)
                        .textInputAutocapitalization(.never)
                        .autocorrectionDisabled()
                        .accessibilityIdentifier("mail_compose_to")
                    if request.mode != .forward {
                        TextField("Cc", text: $cc)
                            .keyboardType(.emailAddress)
                            .textInputAutocapitalization(.never)
                            .autocorrectionDisabled()
                            .accessibilityIdentifier("mail_compose_cc")
                    }
                    TextField("Subject", text: $subject)
                        .accessibilityIdentifier("mail_compose_subject")
                }
                Section {
                    TextEditor(text: $bodyText)
                        .frame(minHeight: 180)
                        .focused($bodyFocused)
                        .accessibilityIdentifier("mail_compose_body")
                } footer: {
                    if let footerText {
                        Text(footerText)
                    }
                }
                if !attachments.isEmpty {
                    Section {
                        ForEach(attachments) { att in
                            HStack(spacing: 8) {
                                Image(systemName: attachmentIcon(mimeType: att.mimeType))
                                    .foregroundStyle(.secondary)
                                Text(att.filename)
                                    .lineLimit(1)
                                Spacer()
                                Button {
                                    attachments.removeAll { $0.id == att.id }
                                } label: {
                                    Image(systemName: "xmark.circle.fill")
                                        .foregroundStyle(.secondary)
                                }
                                .accessibilityIdentifier("mail_compose_attachment_remove")
                            }
                        }
                    }
                }
                if let failure {
                    Section {
                        Text(failure)
                            .foregroundStyle(.red)
                            .accessibilityIdentifier("mail_compose_error")
                    }
                }
            }
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Cancel") { dismiss() }
                        .accessibilityIdentifier("mail_compose_cancel")
                        .disabled(isSending)
                }
                ToolbarItem(placement: .topBarTrailing) {
                    attachButton
                }
                ToolbarItem(placement: .topBarTrailing) {
                    Button {
                        Task { await send() }
                    } label: {
                        if isSending {
                            ProgressView()
                        } else {
                            Image(systemName: "paperplane.fill")
                        }
                    }
                    .accessibilityIdentifier("mail_compose_send")
                    .accessibilityLabel("Send")
                    .disabled(isSending || to.trimmingCharacters(in: .whitespaces).isEmpty
                              || (request.mode == .new && fromAlias == nil))
                }
            }
            .onAppear { bodyFocused = true }
            .interactiveDismissDisabled(isSending)
            .photosPicker(isPresented: $showPhotoPicker, selection: $photoSelection,
                          maxSelectionCount: max(0, Self.maxAttachments - attachments.count), matching: .images)
            .onChange(of: photoSelection) { items in
                guard !items.isEmpty else { return }
                let picked = items
                photoSelection = []
                Task { for item in picked { await loadPhoto(item) } }
            }
            .fileImporter(isPresented: $showFileImporter,
                          allowedContentTypes: Self.documentTypes,
                          allowsMultipleSelection: true) { result in
                loadFiles(result)
            }
        }
    }

    private var attachButton: some View {
        Menu {
            Button {
                showPhotoPicker = true
            } label: {
                Label("Photo Library", systemImage: "photo")
            }
            Button {
                showFileImporter = true
            } label: {
                Label("Files", systemImage: "doc")
            }
        } label: {
            Image(systemName: "paperclip")
        }
        .disabled(isSending || attachments.count >= Self.maxAttachments)
        .accessibilityIdentifier("mail_compose_attach")
    }

    // MARK: Picking

    /// Load a picked photo. HEIC/other non-server-types are transcoded to JPEG so
    /// the server's byte-detected type lands in its allowed set.
    private func loadPhoto(_ item: PhotosPickerItem) async {
        guard let data = try? await item.loadTransferable(type: Data.self) else { return }
        let type = item.supportedContentTypes.first
        if let type, Self.directImageTypes.contains(type.identifier) {
            let ext = type.preferredFilenameExtension ?? "img"
            let mime = type.preferredMIMEType ?? "application/octet-stream"
            addAttachment(MailOutgoingAttachment(filename: "photo.\(ext)", mimeType: mime, data: data))
        } else if let image = UIImage(data: data), let jpeg = image.jpegData(compressionQuality: 0.9) {
            addAttachment(MailOutgoingAttachment(filename: "photo.jpg", mimeType: "image/jpeg", data: jpeg))
        }
    }

    private func loadFiles(_ result: Result<[URL], Error>) {
        guard case .success(let urls) = result else { return }
        for url in urls {
            let scoped = url.startAccessingSecurityScopedResource()
            defer { if scoped { url.stopAccessingSecurityScopedResource() } }
            guard let data = try? Data(contentsOf: url) else { continue }
            let mime = UTType(filenameExtension: url.pathExtension)?.preferredMIMEType
                ?? "application/octet-stream"
            addAttachment(MailOutgoingAttachment(filename: url.lastPathComponent, mimeType: mime, data: data))
        }
    }

    /// Client-side preflight only (fast, friendly failure); the server remains
    /// the authority and re-validates every file and the running total.
    private func addAttachment(_ att: MailOutgoingAttachment) {
        guard attachments.count < Self.maxAttachments else {
            failure = "Up to \(Self.maxAttachments) attachments per message."
            return
        }
        guard att.data.count <= Self.maxAttachmentBytes else {
            failure = "\"\(att.filename)\" is larger than the per-file limit."
            return
        }
        let total = attachments.reduce(0) { $0 + $1.data.count } + att.data.count
        guard total <= Self.maxTotalBytes else {
            failure = "The attachments exceed the total size limit."
            return
        }
        attachments.append(att)
    }

    private func attachmentIcon(mimeType: String) -> String {
        if mimeType.hasPrefix("image/") { return "photo" }
        if mimeType == "application/pdf" { return "doc.richtext" }
        if mimeType.hasPrefix("text/") || mimeType.contains("json") || mimeType.contains("csv") {
            return "doc.text"
        }
        return "doc"
    }

    private func send() async {
        isSending = true
        failure = nil
        do {
            try await api.send(
                mode: request.mode,
                sourceID: request.source?.id,
                aliasID: request.mode == .new ? fromAlias : nil,
                to: to,
                cc: cc,
                subject: subject,
                body: bodyText,
                attachments: attachments
            )
            isSending = false
            dismiss()
            onSent()
        } catch {
            isSending = false
            failure = (error as? JoineryAPIError)?.displayMessage ?? error.localizedDescription
        }
    }
}

File: android/joinery-android-mail/src/main/java/com/getjoinery/mail/ComposeSheet.kt
package com.getjoinery.mail

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.outlined.AttachFile
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import com.getjoinery.android.JoineryApiError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Reply / reply-all / forward / new-message compose. Deliberately lean: the
 * server is the authority on quoting, subject normalization (Re:/Fwd:),
 * threading headers, and the sending identity (for a new message, the picked
 * mailbox) — this sheet collects recipients and the new text. A forward
 * re-attaches the original server-side; new uploads (any mode) attach here
 * via the system picker, mirroring ComposeSheet.swift.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ComposeSheet(
    api: MailApi,
    request: ComposeRequest,
    mailboxes: List<Mailbox>,
    preselectedAlias: Int?,
    onDismiss: () -> Unit,
    onSent: () -> Unit,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val source = request.source

    var to by remember {
        mutableStateOf(
            when (request.mode) {
                ComposeMode.REPLY, ComposeMode.REPLY_ALL ->
                    // Replying to your own outbound message goes back to its
                    // recipient; otherwise to the sender.
                    source?.let { MailDisplay.address(if (it.isOutbound) it.recipient else it.sender) } ?: ""
                else -> ""
            },
        )
    }
    var cc by remember { mutableStateOf("") }
    var subject by remember {
        mutableStateOf(
            when (request.mode) {
                ComposeMode.REPLY, ComposeMode.REPLY_ALL -> prefixed(source?.subject ?: "", "Re:")
                ComposeMode.FORWARD -> prefixed(source?.subject ?: "", "Fwd:")
                ComposeMode.NEW -> ""
            },
        )
    }
    var fromAlias by remember {
        mutableStateOf(
            if (request.mode == ComposeMode.NEW) {
                preselectedAlias ?: mailboxes.firstOrNull()?.aliasId
            } else null,
        )
    }
    var bodyText by remember { mutableStateOf("") }
    val attachments = remember { mutableStateListOf<MailOutgoingAttachment>() }
    var isSending by remember { mutableStateOf(false) }
    var failure by remember { mutableStateOf<String?>(null) }

    fun addPicked(uris: List<Uri>) {
        scope.launch {
            for (uri in uris) {
                val picked = withContext(Dispatchers.IO) { readAttachment(context, uri) } ?: continue
                failure = preflight(attachments, picked)
                if (failure == null) attachments.add(picked)
            }
        }
    }

    val filePicker = rememberLauncherForActivityResult(
        ActivityResultContracts.GetMultipleContents(),
    ) { uris -> addPicked(uris) }

    val title = when (request.mode) {
        ComposeMode.REPLY -> "Reply"
        ComposeMode.REPLY_ALL -> "Reply all"
        ComposeMode.FORWARD -> "Forward"
        ComposeMode.NEW -> "New message"
    }
    val footerText = when (request.mode) {
        ComposeMode.FORWARD -> "The forwarded message and its attachments are included below your text."
        ComposeMode.NEW -> null
        else -> "The original message is quoted below your text."
    }
    val canSend = !isSending && to.trim().isNotEmpty() &&
        !(request.mode == ComposeMode.NEW && fromAlias == null)

    fun send() {
        isSending = true
        failure = null
        scope.launch {
            try {
                api.send(
                    mode = request.mode,
                    sourceId = source?.id,
                    aliasId = if (request.mode == ComposeMode.NEW) fromAlias else null,
                    to = to,
                    cc = cc,
                    subject = subject,
                    body = bodyText,
                    attachments = attachments.toList(),
                )
                isSending = false
                onSent()
            } catch (e: Exception) {
                isSending = false
                failure = (e as? JoineryApiError)?.displayMessage ?: (e.message ?: "Send failed.")
            }
        }
    }

    Dialog(
        onDismissRequest = { if (!isSending) onDismiss() },
        properties = DialogProperties(usePlatformDefaultWidth = false, decorFitsSystemWindows = true),
    ) {
        Surface(Modifier.fillMaxSize()) {
            Scaffold(topBar = {
                TopAppBar(
                    title = { Text(title, maxLines = 1) },
                    navigationIcon = {
                        IconButton(
                            onClick = onDismiss,
                            enabled = !isSending,
                            modifier = Modifier.testTag("mail_compose_cancel"),
                        ) {
                            Icon(Icons.Outlined.Close, contentDescription = "Cancel")
                        }
                    },
                    actions = {
                        AttachMenuButton(
                            enabled = !isSending && attachments.size < MAX_ATTACHMENTS,
                            onPickPhotos = { filePicker.launch("image/*") },
                            onPickFiles = { filePicker.launch("*/*") },
                        )
                        IconButton(
                            onClick = { send() },
                            enabled = canSend,
                            modifier = Modifier.testTag("mail_compose_send"),
                        ) {
                            if (isSending) {
                                CircularProgressIndicator(Modifier.size(20.dp), strokeWidth = 2.dp)
                            } else {
                                Icon(Icons.AutoMirrored.Filled.Send, contentDescription = "Send")
                            }
                        }
                    },
                )
            }) { padding ->
                Column(
                    Modifier
                        .fillMaxSize()
                        .padding(padding)
                        .verticalScroll(rememberScrollState())
                        .padding(horizontal = 16.dp, vertical = 8.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    if (request.mode == ComposeMode.NEW) {
                        FromPicker(mailboxes, fromAlias) { fromAlias = it }
                    }
                    OutlinedTextField(
                        value = to,
                        onValueChange = { to = it },
                        label = { Text("To") },
                        singleLine = true,
                        keyboardOptions = KeyboardOptions(
                            keyboardType = KeyboardType.Email,
                            capitalization = KeyboardCapitalization.None,
                            autoCorrect = false,
                        ),
                        modifier = Modifier.fillMaxWidth().testTag("mail_compose_to"),
                    )
                    if (request.mode != ComposeMode.FORWARD) {
                        OutlinedTextField(
                            value = cc,
                            onValueChange = { cc = it },
                            label = { Text("Cc") },
                            singleLine = true,
                            keyboardOptions = KeyboardOptions(
                                keyboardType = KeyboardType.Email,
                                capitalization = KeyboardCapitalization.None,
                                autoCorrect = false,
                            ),
                            modifier = Modifier.fillMaxWidth().testTag("mail_compose_cc"),
                        )
                    }
                    OutlinedTextField(
                        value = subject,
                        onValueChange = { subject = it },
                        label = { Text("Subject") },
                        singleLine = true,
                        modifier = Modifier.fillMaxWidth().testTag("mail_compose_subject"),
                    )
                    OutlinedTextField(
                        value = bodyText,
                        onValueChange = { bodyText = it },
                        label = { Text("Message") },
                        modifier = Modifier
                            .fillMaxWidth()
                            .heightIn(min = 180.dp)
                            .testTag("mail_compose_body"),
                    )
                    footerText?.let {
                        Text(
                            it,
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    attachments.forEach { att ->
                        Row(
                            Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Icon(
                                if (att.mimeType.startsWith("image/")) Icons.Outlined.Image
                                else Icons.Outlined.Description,
                                contentDescription = null,
                                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                            Text(
                                att.filename,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                modifier = Modifier.weight(1f),
                            )
                            IconButton(
                                onClick = { attachments.removeAll { it.id == att.id } },
                                modifier = Modifier.testTag("mail_compose_attachment_remove"),
                            ) {
                                Icon(Icons.Outlined.Close, contentDescription = "Remove attachment")
                            }
                        }
                    }
                    failure?.let {
                        Text(
                            it,
                            color = MaterialTheme.colorScheme.error,
                            modifier = Modifier.testTag("mail_compose_error"),
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun AttachMenuButton(
    enabled: Boolean,
    onPickPhotos: () -> Unit,
    onPickFiles: () -> Unit,
) {
    var open by remember { mutableStateOf(false) }
    IconButton(
        onClick = { open = true },
        enabled = enabled,
        modifier = Modifier.testTag("mail_compose_attach"),
    ) {
        Icon(Icons.Outlined.AttachFile, contentDescription = "Attach")
    }
    DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
        DropdownMenuItem(
            text = { Text("Photos") },
            leadingIcon = { Icon(Icons.Outlined.Image, contentDescription = null) },
            onClick = { open = false; onPickPhotos() },
        )
        DropdownMenuItem(
            text = { Text("Files") },
            leadingIcon = { Icon(Icons.Outlined.Description, contentDescription = null) },
            onClick = { open = false; onPickFiles() },
        )
    }
}

/** The sending identity for a new message — a picker over the granted
 *  mailboxes (reply/forward keep their implicit source-derived identity). */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun FromPicker(
    mailboxes: List<Mailbox>,
    selected: Int?,
    onSelect: (Int) -> Unit,
) {
    var open by remember { mutableStateOf(false) }
    val current = mailboxes.firstOrNull { it.aliasId == selected }

    ExposedDropdownMenuBox(expanded = open, onExpandedChange = { open = it }) {
        OutlinedTextField(
            value = current?.address ?: "",
            onValueChange = {},
            readOnly = true,
            label = { Text("From") },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = open) },
            modifier = Modifier.fillMaxWidth().menuAnchor().testTag("mail_compose_from"),
        )
        ExposedDropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            mailboxes.forEach { box ->
                DropdownMenuItem(
                    text = { Text(box.address) },
                    onClick = {
                        open = false
                        onSelect(box.aliasId)
                    },
                )
            }
        }
    }
}

// MARK: - Picking

/** Preflight only — mirrors the server's real caps
 *  (`MailboxSender::MAX_UPLOAD_FILES/MAX_UPLOAD_BYTES/MAX_TOTAL_BYTES`); the
 *  server remains the authority and re-validates every file and the total. */
private const val MAX_ATTACHMENTS = 10
private const val MAX_ATTACHMENT_BYTES = 10_485_760
private const val MAX_TOTAL_BYTES = 26_214_400

private fun preflight(
    attachments: List<MailOutgoingAttachment>,
    candidate: MailOutgoingAttachment,
): String? {
    if (attachments.size >= MAX_ATTACHMENTS) {
        return "Up to $MAX_ATTACHMENTS attachments per message."
    }
    if (candidate.data.size > MAX_ATTACHMENT_BYTES) {
        return "\"${candidate.filename}\" is larger than the per-file limit."
    }
    val total = attachments.sumOf { it.data.size } + candidate.data.size
    if (total > MAX_TOTAL_BYTES) {
        return "The attachments exceed the total size limit."
    }
    return null
}

/** Resolve a picked content Uri to bytes + display name + type. The server
 *  re-detects the type from bytes; email carries arbitrary file types, so
 *  nothing is transcoded or filtered here. */
private fun readAttachment(context: Context, uri: Uri): MailOutgoingAttachment? {
    return try {
        val data = context.contentResolver.openInputStream(uri)?.use { it.readBytes() } ?: return null
        var name = uri.lastPathSegment ?: "attachment"
        context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            if (index >= 0 && cursor.moveToFirst()) {
                cursor.getString(index)?.let { name = it }
            }
        }
        val mime = context.contentResolver.getType(uri) ?: "application/octet-stream"
        MailOutgoingAttachment(filename = name, mimeType = mime, data = data)
    } catch (e: Exception) {
        null
    }
}

private fun prefixed(subject: String, prefix: String): String {
    val trimmed = subject.trim()
    if (trimmed.lowercase().startsWith(prefix.lowercase())) return trimmed
    return "$prefix $trimmed"
}
